### ISSUE TITLE
Handle node bootstrapping and readiness for KubeVirt

### ISSUE THREAD

**Issue by rmohr** [kubevirt/kubevirt#314]
In order to only shedule VMs to nodes which are ready to start VMs, we have to think about bootstrapping and  readiness checks for virt-handler. To achieve this, we can think about using taints, tolerances and labels.

Roughly, we would never schedule a VM to a node, which is not marked as ready to start VMs. This involves, that by default all nodes are ignored. Only after virt-handler is ready, we can mark the node as ready for VMs. This also means that we have to think about readiness reporting for virt-handler, so that we can detect missing check-ins from virt-handler, to mark nodes as no longer suitable for us, if virt-handler goes down.

**Comment by michalskrivanek** [comment]
hm, but the suitability is VM specific....so do we just need a set of labels to be announced to scheduler? The first mandatory one being a general readiness  of virt-handler?

**Comment by rmohr** [comment]
@michalskrivanek labels can be a start. It is just not sufficient. If virt-handler is killed, or reports a condition, which makes it impossible to start VMs at all (e.g. no /dev/kvm present), we have to detect that in a cluster component, which removes the readiness from the nodes.

**Comment by michalskrivanek** [comment]
why a cluster component? You mean a cluster service to watch over all nodes and their state and set labels/taints/tolerances accordingly?
I'm not sure...seems to me it's better to delegate that to each node (e.g. a small capability reporting service), and just bubble that info up the stack. Some via metrics (usage, overcomit, etc.) and some via labels (or taints&tolerances, I guess it's just an improvement of general "label", we can use those later to make more intelligent/granular decisions using the stock scheduler)

but maybe that's what you've meant

**Comment by rmohr** [comment]
I don't think that a health monitor, separate to virt-handler can solve this. Then the reachability problem is just moved to the next component on the node. There must be something in the cluster which has the same availability like all other KubeVirt controllers. 

We need something active or passive in the cluster, which then decides if a node is healthy or not.

Kubernetes has this on nodes:

```
    conditions:
    - lastHeartbeatTime: 2017-07-14T10:31:11Z
      lastTransitionTime: 2017-07-14T07:23:06Z
      message: kubelet is posting ready status
      reason: KubeletReady
      status: "True"
      type: Ready
```

FYI, quoting from [Kubernets Docs](https://kubernetes.io/docs/concepts/architecture/nodes/#condition). The Ready condition is

> [ ...] True if the node is healthy and ready to accept pods, False if the node is not healthy and is not accepting pods, and Unknown if the node controller has not heard from the node in the last 40 seconds

**Comment by michalskrivanek** [comment]
why would we need more than that? One is the "control plane", the infrastructure health, for which the above should be sufficient ("ready" corresponding to Up host state in oVirt, "not ready" corresponding to NonOperational, and "unknown" being NotResponding). It doesn't tell you anything about the ability to run VMs, that's scheduler's job. But for the whole management system to function you don't really need more than that.

**Comment by rmohr** [comment]
@michalskrivanek this ready just means that you can schedule pods. This does not mean that our kubevirt infra is present on the host. Without that, the VM will never come up. With a label in the scheduler you can't completely solve that problem, because our KubeVirt infra on the node can be removed, not reachable or malfunction. Finally the node can miss some virtualization capabilities, which KubeVirt in general needs.

With an extra condition, like "Ready" condition, which the scheduler can take into account. Further, we would pollute important spec fields like labels with less internal stuff.

**Comment by michalskrivanek** [comment]
do not mix up infra with capabilities. Those need to be clearly separated. 
You may only need a check for "kubevirt infra" (virt-handler) functioning correctly on the node - but even that might not really be needed if virt-launcher errors out correctly. There is going to be a higher-level handling in scheduler to then flag such host as unusable/problematic

**Comment by rmohr** [comment]
> do not mix up infra with capabilities. Those need to be clearly separated.

Ok, so with /dev/kvm I mean "required infra" then.

> You may only need a check for "kubevirt infra" (virt-handler) functioning correctly on the node - but even that might not really be needed if virt-launcher errors out correctly. There is going to be a higher-level handling in scheduler to then flag such host as unusable/problematic

What I explained above, is how kubernetes finds out if a node is ready. So if the kubelet does not report in (on the node conditions) that it is ready, the node controller (the higher-level) will mark the node according to the quote from the kubernetes docs. The scheduler can then take the conditions into account. I would want to follow that. IMHO, if virt-handler errors out correctly, or not does not really have anything to do with that.

**Comment by michalskrivanek** [comment]
ah, so no, I do not consider "/dev/kvm" a required infra. We can't make such assumptions as the required "infra" for running a certain VM is much more detailed than that. Even a lack of /dev/kvm might be a valid state for full emulation for trying things out.
I believe the basic concept of node being ready communication-wise is good enough, everything else is scheduler's responsibility.

note I don't mean virt-handler to error out, I've meant the launch process, the virt-launcher/libvirt/qemu process. In case of the scheduler scheduling a POD on a node where it doesn't run (we can't avoid those cases) there needs to be a response feedback in the scheduler to stop doing that repeatedly, e.g. by flagging such node as unsuitable.

**Comment by rmohr** [comment]
> ah, so no, I do not consider "/dev/kvm" a required infra. We can't make such assumptions as the required "infra" for running a certain VM is much more detailed than that. Even a lack of /dev/kvm might be a valid state for full emulation for trying things out.
> I believe the basic concept of node being ready communication-wise is good enough, everything else is scheduler's responsibility.

Then let's take connectivity to libvirt. Maybe that is more what you would want to hear in this context. Just something like this. If virt-handler can't connect to libvirt, the node is not ready for VMs. So virt-handler would mark the node as not ready for VMs. virt-handler would be e.g. responsible for `Ready` and `NotReady`, but it can't be responsible for `Unknown`. There the cluster has to help out.

> note I don't mean virt-handler to error out, I've meant the launch process, the virt-launcher/libvirt/qemu process. In case of the scheduler scheduling a POD on a node where it doesn't run (we can't avoid those cases) there needs to be a response feedback in the scheduler to stop doing that repeatedly, e.g. by flagging such node as unsuitable.

First, this issue I create really is just about the infra readiness. Second, what you describe here is of course right, but not necessarily related to the infra readiness. So while agreeing with you, it is not the topic of this issue. Does that work for you?

**Comment by michalskrivanek** [comment]
> On 14 Jul 2017, at 15:26, Roman Mohr <notifications@github.com> wrote:
> 
> ah, so no, I do not consider "/dev/kvm" a required infra. We can't make such assumptions as the required "infra" for running a certain VM is much more detailed than that. Even a lack of /dev/kvm might be a valid state for full emulation for trying things out.
> I believe the basic concept of node being ready communication-wise is good enough, everything else is scheduler's responsibility.
> 
> Then let's take connectivity to libvirt. Maybe that is more what you would want to hear in this context. Just something like this. If virt-handler can't connect to libvirt, the node is not ready for VMs. So virt-handler would mark the node as not ready for VMs. virt-handler would be e.g. responsible for Ready and NotReady, but it can't be responsible for Unknown. There the cluster has to help out.
> 
> 
no matter which aspect you choose, it means “ready” for one VM and “not ready” for another VM. Yes, there’s going to also be a “not ready for any VM whatsoever” which the node itself can decide, but I do not see useful to cover few cases (like virt-handler can’t connect to libvirt) in a special way.

**Comment by rmohr** [comment]
> there’s going to also be a “not ready for any VM whatsoever” which the node itself can decide

The issue only tries to deal with that.

> but I do not see useful to cover few cases (like virt-handler can’t connect to libvirt) in a special way.

What exactly virt-handler will consider when setting ready or not, is not so imortant, the libvirt example is something which might be worth to be considered by virt-handler, but really this is just about, the node is ready in general, not ready, or in an unknown state in the context of KubeVirt.

**Comment by fabiand** [comment]
@rmohr could it be that this is done?

**Comment by rmohr** [comment]
Yes. Implemented in #838
